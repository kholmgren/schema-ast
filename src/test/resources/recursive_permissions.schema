definition tenant/folder {
	relation parent: tenant/folder
	relation reader: tenant/user

	// Note that since `parent` refers to `tenant/folder` (which is this type), `parent->read` will call
	// this same permission, therefore following `read` for *that tenant/folder's* parent.
	permission read = reader + parent->read
}

definition tenant/user {}