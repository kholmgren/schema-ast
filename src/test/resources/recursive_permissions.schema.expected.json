{
  "@type" : "schema",
  "definitions" : [ {
    "@type" : "definition",
    "name" : "tenant/folder",
    "relations" : [ {
      "@type" : "relation",
      "name" : "parent",
      "expression" : {
        "@type" : "ref",
        "ref" : "tenant/folder"
      }
    }, {
      "@type" : "relation",
      "name" : "reader",
      "expression" : {
        "@type" : "ref",
        "ref" : "tenant/user"
      }
    } ],
    "permissions" : [ {
      "@type" : "permission",
      "name" : "read",
      "expression" : {
        "@type" : "union",
        "left" : {
          "@type" : "ref",
          "ref" : "reader"
        },
        "right" : {
          "@type" : "walk",
          "refs" : [ "parent", "read" ]
        }
      }
    } ]
  }, {
    "@type" : "definition",
    "name" : "tenant/user"
  } ]
}